File: src/services/inventory/get-product-stock.ts
import { apiClient } from '@/shared/lib/api/client';
import { BASE_PATH } from './_base';

export interface GetProductStockParams {
  productVariationId?: string;
}

export interface GetProductStockResponse {
  productId: string;
  productVariationId: string | null;
  totalQuantity: number;
}

/**
 * GET /products/{productId}/stock — rolls up every variation/warehouse unless
 * `productVariationId` narrows it to one variation. Backs the same query the
 * Order Service's gRPC stock lookup uses.
 */
export function getProductStock(
  productId: string,
  params: GetProductStockParams = {},
): Promise<GetProductStockResponse> {
  return apiClient
    .get(`${BASE_PATH}/products/${productId}/stock`, { params })
    .then((res) => res.data);
}

File: src/Services/Inventory/Inventory.API/Endpoints/Inventory/GetProductStock.cs
using SmartEcommerce.BuildingBlock.Application.Abstractions.Common;
using SmartEcommerce.BuildingBlock.Infrastructure.Authorization;
using SmartEcommerce.BuildingBlock.SharedKernel.Constants;
using SmartEcommerce.BuildingBlock.SharedKernel.Extensions;

using SmartEcommerce.Inventory.Application.Features.Inventories.Queries.GetProductStock;

namespace SmartEcommerce.Inventory.API.Endpoints.Inventory;

public sealed class GetProductStockEndpoint : ICarterModule
{
    private readonly string[] API_DESC = [
        "## Get Product Stock",
        "",
        "Returns total stock for a product. ProductId alone rolls up every variation and",
        "warehouse; adding ProductVariationId narrows the rollup to that single variation.",
        "This is the same query the gRPC stock lookup (consumed by Order Service) is backed by.",
        "",
        "### Route Parameters",
        "- **productId**: Unique identifier of the product (required, must be valid GUID)",
        "",
        "### Query Parameters",
        "- **productVariationId**: Optional, narrows the rollup to a single variation",
    ];

    public void AddRoutes(IEndpointRouteBuilder app)
    {
        app.MapGet("/products/{productId}/stock", Handle)
            .WithTags("Inventory")
            .RequireAuthorization(AuthorizationPolicies.RequireAuthenticated)
            .WithName("GetProductStock")
            .WithDisplayName("Get Product Stock API")
            .WithDescription(API_DESC.JoinToString("\n"))
            .Produces<ApiResponse<GetProductStockResponse>>(StatusCodes.Status200OK);
    }

    private static async Task<IResult> Handle(
        [FromRoute] Guid productId,
        [FromQuery] Guid? productVariationId,
        [FromServices] ISender sender,
        CancellationToken ct = default)
    {
        var query = new GetProductStockQuery(productId, productVariationId);
        var response = await sender.Send(query, ct);

        return Results.Ok(ApiResponse<GetProductStockResponse>.Ok(response));
    }
}
